Assets/MixedRealityToolkit/Utilities/AssetScriptReferenceRetargeter.cs

```diff
@@ -0,0 +1,593 @@
+
+        private const string ScriptFileIdConstant = "11500000";
+
+        [MenuItem("Assets/Retarget To DLL")]
+        public static void RetargetAssets()
+        {
+            try
+            {
+                Debug.Log("Starting to Retarget Assets.");
+                RunRetargetToDLL();
+                Debug.Log("Complete.");
+            }
+            catch (Exception ex)
+            {
+                Debug.LogError("Failed.");
+                Debug.LogException(ex);
+            }
+        }
+
+        private static void RunRetargetToDLL()
+        {
+            string[] allFilesUnderAssets = Directory.GetFiles(Application.dataPath, "*", SearchOption.AllDirectories);
+
+            Dictionary<string, ClassInformation> scriptFilesReferences = ProcessScripts(allFilesUnderAssets);
+            Debug.Log($"Found {scriptFilesReferences.Count} script file references.");
+
+            // DLL name to Guid
+            Dictionary<string, string> asmDefMappings = RetrieveAsmDefGuids(allFilesUnderAssets);
+
+            Dictionary<string, ClassInformation> compiledClassReferences = ProcessCompiledDLLs("PackagedAssemblies", Application.dataPath.Replace("Assets", "NuGet/Plugins/EditorPlayer"), asmDefMappings);
+            Debug.Log($"Found {compiledClassReferences.Count} compiled class references.");
+
+            Dictionary<string, Tuple<string, long>> remapDictionary = new Dictionary<string, Tuple<string, long>>();
+
+            foreach (var pair in scriptFilesReferences)
+            {
+                if (compiledClassReferences.TryGetValue(pair.Key, out ClassInformation compiledClassInfo))
+                {
+                    remapDictionary.Add(pair.Value.Guid, new Tuple<string, long>(compiledClassInfo.Guid, compiledClassInfo.FileId));
+                }
+                else
+                {
+                    // Switch to throwing exception later
+                    Debug.LogWarning($"Can't find a compiled version of the script: {pair.Key}; guid: {pair.Value.Guid}");
+                }
+            }
+
+            ProcessYAMLAssets(allFilesUnderAssets, Application.dataPath.Replace("Assets", "NuGet/Content"), remapDictionary, asmDefMappings);
+        }
+
+        private static Dictionary<string, string> RetrieveAsmDefGuids(string[] allFiles)
+        {
+            int lengthOfPrefix = Application.dataPath.IndexOf("Assets");
+            Dictionary<string, string> dllGuids = new Dictionary<string, string>();
+            foreach (string asmdefFile in allFiles.Where(t => t.EndsWith(".asmdef")))
+            {
+                string asmdefText = AssetDatabase.LoadAssetAtPath<AssemblyDefinitionAsset>(asmdefFile.Substring(lengthOfPrefix)).text;
+                string dllName = JsonUtility.FromJson<AssemblyDefinitionStub>(asmdefText).name;
+                string guid = File.ReadAllLines($"{asmdefFile}.meta")[1].Substring(6);
```

Comment:
>string guid = File.ReadAllLines($"{asmdefFile}.meta")[1].Substring(6); [](start = 16, length = 70)

Do you know if this structure of the .meta file will be consistent/solid going forward (i.e. we can take this dependency) or if not, if there's any way we can error out/assert if this isn't actually a valid GUID anymore? i.e. they changed the data structure now and the guid is on line 3 instead of line 2

Reply:
This can only come with Unity upgrades, at which time we will test all this; but I will do a TryParse to validate we did get a guid indeed.

---
In reply to: [277827565](https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/4051#discussion_r277827565) [](ancestors = 277827565)